### ISSUE TITLE
No type hints on nn.Parameter

### ISSUE THREAD

**Issue by theo-m** [pytorch/pytorch#25399]
## 🐛 Bug

Mypy fails on simple calls on `nn.Parameter`

## To Reproduce

1. `conda create -n torch-issue python=3.6 -y`
2. `conda activate torch-issue`
3. `conda install pytorch torchvision -c pytorch -y`
4. `pip install mypy=0.720`
5. Run with `mypy test`
```python
# test.py
import torch

x = torch.zeros(10, 3)
p = torch.nn.Parameter(x)
p = torch.nn.Parameter(data=x)  # second try
```

This yields:
```
test.py:5: error: Too many arguments for "Parameter"
test.py:6: error: Unexpected keyword argument "data" for "Parameter"
```

## Expected behavior

Type check should pass

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.6
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.2.0
[pip] torchvision==0.4.0a0+6b959ee
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      233  
[conda] mkl-service               2.0.2            py36h1de35cc_0  
[conda] mkl_fft                   1.0.14           py36h5e564d8_0  
[conda] mkl_random                1.0.2            py36h27c97d8_0  
[conda] pytorch                   1.2.0                   py3.6_0    pytorch
[conda] torchvision               0.4.0                  py36_cpu    pytorch

## Additional
I've noticed there is no mention to Parameter in the type stub file, maybe that's the cause?

cc @ezyang

**Comment by VitalyFedyunin** [comment]
Feel free to submit PR to fix it

**Comment by ezyang** [comment]
> I've noticed there is no mention to Parameter in the type stub file, maybe that's the cause?

Yes, almost definitely.

**Comment by theo-m** [comment]
> Feel free to submit PR to fix it

I will, I've looked up merged PR with `module: typing` as a label but only two came up so I'm a bit unclear on steps to do it. 

I'm also unclear on whether if the typing code should be in `torch/__init__.pyi.in` or in `torch/nn/__init__.pyi` given that the latter imports the `Parameter` class from the source.

**Comment by ezyang** [comment]
You should generally put typing definitions in the pyi file of whatever module originally defined them. That way, direct access to the module gets typed directly.

For general docs about typing, I find the mypy docs are pretty good: https://mypy.readthedocs.io/en/latest/

**Comment by theo-m** [comment]
Hi, some follow up:

I made some changes to `nn/__init__.pyi` but I can't build from source, hence I can't test locally and I wouldn't push a PR without having tried something.
I'm encoutering issues that I describe in [this comment](https://github.com/pytorch/pytorch/issues/24831#issuecomment-526536609).
I couldn't get an answer there, maybe you can help?

**Comment by ezyang** [comment]
> I wouldn't push a PR without having tried something on my own install.

If you're pretty sure your PR is fine, you can just push and let CI do its thing. Not too sure about your build error; maybe you can uninstall google protobuf from your system and try again.

**Comment by theo-m** [comment]
Well that's the thing I'm not too sure of my changes.

1. the `.pyi` file imports `Parameter`, my understanding is that if I were to type it I'd have to remove that initial line, but not sure, would like to test.
2. Ideally I'd like the typing of `Parameter` to inherit what's been done in the root `.pyi.in` file for the `Tensor` class since `Parameter` is a child of the latter. I can't do that since it would create a cycle, and I'm unclear on how to declare the `Parameter` typing to the root stub file - nor if it's actually a feature since as I understand it we want typing to be done at the module level.

**Comment by ezyang** [comment]
In general, you should replicate the module import structure of the original py files, which definitionally don't have a cycle. It's possible we messed up some imports in the pyi files; in that case, you'd have to fix it.
